README.md
Set of the bash(lyk) scripts for development and build packages.

Prepared debian packages can be found in the repository **ppa:yds/bashlyk** (`sudo add-apt-repository ppa:yds/bashlyk`).


ChangeLog
Sun, 07 May 2023 16:36:42 +0000
  New upstream release
        Version 1.0
Change the content of a "Git" keyword
- represent as a combination of the current git tag (version), the
  number of commits since the creation of the current tag (build), the
  total number of commits before the creation of the current tag (stable
  revision), as well as the timestamp and ID of the author of the
  changes
- improve count of total number of commits
- add pre-commit hook for "Id" and "Git" keywords supporting


Sun, 11 Sep 2022 17:58:30 +0400
  New upstream release
        Version 0.9
kolchan-automake:
- add autoscan control option
- disable autoscanning by default to prevent including unwanted settings
  from external libraries, e.g. node_modules
- generate a hierarchical list of configurations when autoscanning is
  disabled
- improve version and author name parsing
git-add-id:
- add "Git" keyword - combination of the actual tag (version), the total
  number of commits before  setting  the  actual  tag (revision) and the
  number of commits of the actual tag.
- remove lead 'v' from version id
kolchan-*:
- improve usage example
- use git tag first to get the current version
- show config by debug level
*-up2deb:
- fixed debian files handling
debian/control:
- add dependency dh-make, bashlyk
README.md:
- add PPA link for prepared debian packages


Sun, 09 Jun 2019 00:12:24 +0400
  New upstream release
        Version 0.8
- automake*, build-deb*, uupdeb renamed to kolchan-*:
  automake.sh  -> kolchan-automake
  build-deb.sh -> kolchan-builddeb
  uupdeb       -> kolchan-up2deb
- add  CLI  options  and configuration support, build artifacts saved to
  config for use by other tools, improved build automation
- add  libsys.sh - rsync object  implementation,  rsync code replaced by
  SYS::RSYNC object
- add debianization infrastructure
- *-up2deb - select 'dh_make' or 'uupdate' automatically
- improved  robodoc  EXAMPLE  section  - add  reproducible  example from
  scratch
- improve description and usage information
- added simple architecture dependency check:
  *-automake: check 'gcc' keyword from config.log and save boolean value
  to global section with key 'archindep'
  *-up2deb: set dh_make option '-i' if 'archindep' value is 'true'
- *-builddeb: use mode 'binary' by default, 'binary' mode is transformed
  to "all" if the 'archindep' option is true, to 'any' otherwise.


Wed, 08 May 2019 10:48:30 +0400
	New upstream release
        Version 0.7
- added base support for jenkins pipelines


Mon, 01 Apr 2019 22:51:02 +0400
	New upstream release
        Version 0.6
- added new script for autotools use - automake.sh


Mon, 25 Mar 2019 23:55:33 +0400
	New upstream release
        Version 0.5
- added new scripts for debian package building:
  - build-deb.sh - wrapper for dpkg-buildpackage
  - uupdeb - wrapper for uupdate


Mon, 16 Jul 2018 18:24:41 +0400
	New upstream release
        Version 0.4
- git-add-id - added 'user.id' value from git config:
  first try use "user.id" value from git config, if is empty use $USER value


Fri, 17 Feb 2017 15:36:19 +0400
	New upstream release
        Version 0.3
- git-add-with-keywords renamed to git-add-id
- added autotools and robodoc configuration files


Thu, 23 Jun 2016 14:15:53 +0400
	New upstream release
        Version 0.2
- added git-add-with-keywords (early gitadd.sh) - wrapper for emulation
  svn-keyword 'Id' by using calculating total count of commits from stdout of
  'git shortlog -s'
- revision number (count of commits) incremented anticipating the expected commit


Mon, 20 Jun 2016 17:19:32 +0400
	Initial release
        Version 0.1
- added scripts to convert svn to git: subgit.sh, svnlog2rev.sh, svnsetlog.sh


src/gitadd.sh
#!/bin/bash
#
# $Id: gitadd.sh 10 2016-06-23 12:41:05+04:00 toor $
#
. bashlyk
#
#
#
udfMain() {

	udfThrowOnCommandNotFound bc date git grep sed xargs

	local fmt fn ic ts ui

	fmt='sed -i -r -e "s/%sId(.*)?%s/%s/" %s'
	ic=$( git shortlog -s | grep -oP '^\s+\d+\s+' | xargs | tr ' ' '+' | bc )

	eval set -- "$(_ sArg)"

	[[ -d "$1" ]] && eval $( udfOnError exitecho iErrorNonValidArgument "this version do not support directories, only files \($1\)..." )

	[[ -f "$1" && -n "$( git status -s "$1" | grep -P "^[ AM?][M?]\s$1" )" ]] || {

		eval $(udfOnError retwarn iErrorNonValidArgument "$1 is mask, not exist or not modified, but may be indexed...")

	}

	fn="${1##*/}"
	ic=$(( ic + 1 ))
	ts="$( date --rfc-3339=s )"
	ui="$(_ sUser)"

	if udfWarnOnEmptyVariable fn ic ts ui; then

		s="$( printf '\044Id: %s %d %s %s \044\n' "$fn" "$ic" "$ts" "$ui" )"
		eval "$( printf "$fmt" '\\\$' '\\\$' '$s' '$fn')" && git add "$1"

	fi

}
#
#
#
udfMain
#


src/subgit.sh
#!/usr/bin/env bash
#
# $Id: subgit.sh 57 2019-05-08 10:54:57+04:00 yds $
#
_bashlyk=devtools . bashlyk
#
#
#
subgit::main() {

  throw on CommandNotFound subgit

  local pathGIT pathSVN project
  local -a aAuthors

  CFG cfg

  cfg.bind.cli authors{a}:-- config{c}: help{h} git{g}: project{p}: svn{s}:

  cfg.storage.use $( cfg.getopt config )

  cfg.load []git,project,svn [authors]=

  pathGIT=$( cfg.get git )
  pathSVN=$( cfg.get svn )
  project=$( cfg.get project )

  s=$( cfg.get [authors] )
  eval "${s/declare -a a/declare -a aAuthors}"

  cfg.free

  : ${pathGIT:=opt/dat/git}
  : ${pathSVN:=opt/dat/svn}

  if [[ $project ]]; then

    pathGIT+="/${project}.git"
    pathSVN+="/${project}"

  fi

  throw on NoSuchDir $pathSVN

  #
  # prepare authors substitution
  #
  std::temp fnAuthors
  for s in "${aAuthors[@]}"; do

    echo "$s" >> $fnAuthors

  done
  [[ -s $fnAuthors ]] && fnAuthors="--authors-file $fnAuthors" || fnAuthors=

  ## TODO use svn(admin) and git tools for checking repos validity

  subgit import $fnAuthors --svn-url file://${pathSVN} $pathGIT

}
#
#
#
subgit::main
#


src/automake.sh
#!/usr/bin/env bash
#
# $Id: automake.sh 69 2019-05-25 08:32:57+04:00 yds $
#
_bashlyk_log=nouse _bashlyk=devtools . bashlyk
#
#
#
automake::main() {

  throw on NoSuchFile ChangeLog
  throw on CommandNotFound autoscan autoheader aclocal autoconf autoreconf automake cat cut grep mv pwd robodoc head tr touch sed xargs

  local fn sPackage sVersion sAuthor pathWork
  
  #
  # TODO add package name by CLI
  #

  : ${sVersion:=0.1}

  errorify on CommandNotFound git && : ${sAuthor:="$( git config --get user.name ) <$( git config --get user.email )>"}

  : ${sAuthor:="${DEBFULLNAME:-$USER} <${DEBEMAIL:-$USER@localhost.localdomain}>"}

  pathWork="$( pwd )"

  [[ -s AUTHORS ]] || echo "$sAuthor" > AUTHORS

  for fn in NEWS README TODO; do [[ -f $fn ]] || touch $fn; done

  sPackage=${pathWork##*/}
  sPackage=${sPackage/-$sVersion/}
  sVersion=$(grep -i version ChangeLog | head -n 1 | xargs | cut -f 2 -d' ')
  : ${sVersion:=0.001}

  sEmail="$(grep -o -E '<.*>' AUTHORS | tr -d '<|>' | head -n 1)"

  autoscan
  mv configure.scan configure.ac || throw on NoSuchFile configure.ac

  autoheader
  sed -i -e "s/AC_INIT.*/AC_INIT(${sPackage}, ${sVersion}, ${sEmail})\nAM_INIT_AUTOMAKE/ig" configure.ac

  aclocal
  autoconf
  autoreconf
  automake --add-missing --copy

  errorify on CommandNotFound git-add-id && git-add-id

  throw on CommandNotFound ./configure make && ./configure --prefix=/usr && make

}
#
#
#
automake::main
#


src/build-deb.sh
#!/usr/bin/env bash
#
# $Id: build-deb.sh 70 2019-05-25 10:15:32+04:00 yds $
#
_bashlyk_log=nouse _bashlyk=kolchan . bashlyk

: ${_kolchan_pathLib:=/usr/share/kolchan}

. ${_kolchan_pathLib}/libsys.sh || throw

#
#
#
: ${DEBUGLEVEL:=2}

buildpackage::main() {

  throw on CommandNotFound dpkg-buildpackage head grep rsync sed tail wc

  local fn fnConfig pathTarget pathSource sCodeName sMode sProject
  local -a

  CFG cfg
  cfg.bind.cli mode{m}: path-source{s}: path-target{t}: config{c}:
  fnConfig=$( cfg.getopt config ) || fnConfig=$( cfg.storage.use && cfg.storage.show )
  cfg.storage.use $fnConfig
  err::debug 0 project configuration $( cfg.storage.show )
  cfg.load

       sMode="$( cfg.get     []mode    )" ||      sMode='binary'
    sProject="$( cfg.get     []project )" ||   sProject=$PROJECT
  pathSource="$( cfg.get [path]source  )" || pathSource=~/src

  if ! pathTarget="$( cfg.get [path]target )"; then

    if [[ $PRJ_BUILDINFO && -s $PRJ_BUILDINFO ]]; then

      pathTarget="$( head -n 1 $PRJ_BUILDINFO )"

    else

      pathTarget="$( pwd )"

    fi

  fi

  throw on NoSuchDir $pathTarget

  cd $pathTarget || error NotPermitted throw -- $pathTarget

  if [[ -d debian.upstream && -d debian ]]; then

    SYS::RSYNC rsync
    rsync.title      = 'merge debian.upstream with debian'
    rsync.options    = -arv --exclude changelog
    rsync.pathSource = debian.upstream/
    rsync.pathTarget = debian/
    rsync.onFailure  = throw
    rsync.onSuccess  = rm -r debian.upstream
    rsync.run
    rsync.free

  fi

  throw on NoSuchFile debian/changelog

  if head -n 1 debian/changelog | grep -q UNRELEASED; then

    CFG cfgLSB
    cfgLSB.storage.use /etc/lsb-release
    cfgLSB.load []DISTRIB_CODENAME

    if sCodeName="$( cfgLSB.get []DISTRIB_CODENAME )"; then

      err::debug 2 debian/changelog - use codename $sCodeName
      sed -i "1 s/UNRELEASED/$sCodeName/" debian/changelog

    else

      err::debug 2 DISTRIB_CODENAME value not loaded from /etc/lsb-release

    fi

    cfgLSB.free

  else

    err::debug 2 UNRELEASED tag not found on debian/changelog

  fi

  #
  # TODO split buildinfo to the 'binary' and 'source' sections (or 'mode' option)
  #

  std::temp fn
  echo -n "start building:"
  while read; do

  if   [[ $REPLY =~ dpkg-genchanges.--build=(source|any|all|binary|full).\>../(.*)$ ]]; then
    cfg.set [$sMode]${BASH_REMATCH[1]} = ${BASH_REMATCH[2]}
  elif [[ $REPLY =~ dpkg-deb:.building.package.*in.*../(.*\.deb) ]]; then
    cfg.set [$sMode]package = ${BASH_REMATCH[1]}
  elif [[ $REPLY =~ ^return.code:.(.*)$ ]]; then
    cfg.set [$sMode]status = ${BASH_REMATCH[1]}
    if [[ $( cfg.get [$sMode]status ) != '0' ]]; then
      echo "fail.."
      err::debug 0 stderr:
      cat $fn 1>&2
    else
      echo "ok!"
    fi
    cfg.set [$sMode]timestamp = $( std::date %s )
  else
    echo $REPLY >> $fn
    echo -n '.'
  fi

  done< <( LC_ALL=C dpkg-buildpackage --build=$sMode -rfakeroot 2>&1; echo "return code: $?" )

  err::debug 0 build states:
  cfg.show
  cfg.save
  cfg.free

}
#
#
#
buildpackage::main
#


src/svnlog2revs.sh
#!/usr/bin/env bash
#
# $Id: svnlog2revs.sh 57 2019-05-08 10:54:57+04:00 yds $
#
_bashlyk=devtools . bashlyk
#
#
#
svnlog2revs::main() {

  throw on CommandNotFound rm sed touch
  throw on MissingArgument $1
  throw on NoSuchFile $1

  eval set -- "$(_ sArg)"

  local s fn fnRev tsRev

  fn=$1

  sed -i -r -e "s/\t/%SVNLOG2REVSTABULA%/g" -e "s/^(\s+)/%SVNLOG2REVSIDENT%\1/" $fn

  while read s; do

    [[ "$s" =~ ^------------------------------------------------------------------------$ ]] && continue
    if [[ "$s" =~ ^(r[0-9]+)[[:space:]]\|[[:space:]][[:graph:]]+[[:space:]]\|[[:space:]](.*)[[:space:]]\(.*\).*[0-9]+.line(s)?$ ]]; then

      if [[ $fnRev && -f $fnRev && $tsRev ]]; then

        [[ "$2" == "add-svn-rev" ]] && printf -- "--\n--\t(svn rev %s)\n" "$fnRev" >> $fnRev
        sed -i -r -e "s/%SVNLOG2REVSTABULA%/\t/g" -e "s/%SVNLOG2REVSIDENT%//" $fnRev
        touch --date="$tsRev" $fnRev

      fi

      if [[ ${BASH_REMATCH[1]} ]]; then

        fnRev="${BASH_REMATCH[1]}"
        tsRev="${BASH_REMATCH[2]}"
        rm -f $fnRev
        echo "$fnRev $tsRev"

      else

        error Unexpected throw "$s"

      fi

      read s
      continue

    fi

    [[ $fnRev =~ ^r[0-9]+$ ]] || continue
    echo "$s" >> $fnRev

  done < $fn

  if [[ $fnRev && -f "$fnRev" && $tsRev ]]; then

    [[ $2 == add-svn-rev ]] && printf -- "--\n--\t(svn rev %s)\n" "$fnRev" >> $fnRev
    sed -i -r -e "s/%SVNLOG2REVSTABULA%/\t/g" -e "s/%SVNLOG2REVSIDENT%//" $fnRev
    touch --date="$tsRev" $fnRev

  fi

  sed -i -r -e "s/%SVNLOG2REVSTABULA%/\t/g" -e "s/%SVNLOG2REVSIDENT%//" $fn

}
#
#
#
svnlog2revs::main
#


src/libsys.sh
#
# $Id: libsys.sh 89 2019-05-29 14:18:49+04:00 yds $
#
#****h* kolchan/libsys
#  DESCRIPTION
#    OOP style wrappers for various system commands:
#    SYS::RSYNC - support for rsync
#  USES
#    libcfg
#  EXAMPLE
#    # create instance from SYS::RSYNC class
#    SYS::RSYNC rsync
#    # set various properties for rsync operation
#    rsync.pathSource = /tmp/source
#    rsync.pathTarget = /tmp/target
#    rsync.options    = -aCrv --delete-after
#    rsync.debugLevel = 2
#    rsync.title      = sync $( rsync.pathSource) to $( rsync.pathTarget )
#    rsync.onSuccess  = rm -r $( rsync.pathSource)
#    # start rsync
#    rsync.run
#    # destroy SYS::RSYNC object, free resources
#    rsync.free
#  AUTHOR
#    Damir Sh. Yakupov <yds@bk.ru>
#******
#***iV* libsys/bash compatibility
#  DESCRIPTION
#    Compatibility checked by bashlyk (bash version 4.xx or more required)
#    $_BASHLYK_LIBSYS provides protection against re-using of this module
#  SOURCE
: ${_bashlyk_pathLib:="/usr/share/bashlyk"}
[ -n "$_BASHLYK_LIBSYS_RSYNC" ] && return 0 || _BASHLYK_LIBSYS_RSYNC=1
[ -n "$_BASHLYK" ] || . ${_bashlyk_pathLib}/bashlyk || eval '                  \
                                                                               \
    echo "[!] bashlyk loader required for ${0}, abort.."; exit 255             \
                                                                               \
'
#******
#****L* libsys/Used libraries
# DESCRIPTION
#   Loading external libraries
# SOURCE
[[ -s ${_bashlyk_pathLib}/libcfg.sh ]] && . "${_bashlyk_pathLib}/libcfg.sh"
#******
#****G* libsys/Global Variables
#  DESCRIPTION
#    Global variables of the library
#  SOURCE
# rsync error states definition
 _rsync_hError[0]="Success"
 _rsync_hError[1]="Syntax or usage error"
 _rsync_hError[2]="Protocol incompatibility"
 _rsync_hError[3]="Errors selecting input/output files, dirs"
 _rsync_hError[4]="Requested action not supported: an attempt was made to manipulate 64-bit files on a platform that cannot support them; or an option was specified that is supported by the client and not by the server."
 _rsync_hError[5]="Error starting client-server protocol"
 _rsync_hError[6]="Daemon unable to append to log-file"
_rsync_hError[10]="Error in socket I/O"
_rsync_hError[11]="Error in file I/O"
_rsync_hError[12]="Error in rsync protocol data stream"
_rsync_hError[13]="Errors with program diagnostics"
_rsync_hError[14]="Error in IPC code"
_rsync_hError[20]="Received SIGUSR1 or SIGINT"
_rsync_hError[21]="Some error returned by waitpid()"
_rsync_hError[22]="Error allocating core memory buffers"
_rsync_hError[23]="Partial transfer due to error"
_rsync_hError[24]="Partial transfer due to vanished source files"
_rsync_hError[25]="The --max-delete limit stopped deletions"
_rsync_hError[30]="Timeout in data send/receive"
_rsync_hError[35]="Timeout waiting for daemon connection"
#
_bashlyk_iErrorMissingMethod=165
_bashlyk_iErrorBadMethod=164
_bashlyk_hError[$_bashlyk_iErrorMissingMethod]="instance failed - missing method"
_bashlyk_hError[$_bashlyk_iErrorBadMethod]="instance failed - bad method"
#
declare -rg _bashlyk_settings_sys_rsync='
    debugLevel fileLog onFailure onSuccess options pathSource pathTarget title truncateLog
'
declare -rg _bashlyk_methods_sys_rsync='
    run free
'
#******
#****e* libsys/SYS::RSYNC
#  SYNOPSIS
#    SYS::RSYNC [<id>]
#  DESCRIPTION
#    constructor for new instance <id> of the SYS::RSYNC "class" (object)
#  NOTES
#    constructor
#  ARGUMENTS
#    valid variable name for created instance, default - used class name RSYNC as
#    instance
#  ERRORS
#    InvalidVariable  - invalid variable name for instance
#    IniMissingMethod - method not found
#    IniBadMethod     - bad method
#  EXAMPLE
#    SYS::RSYNC tnew                                                           #? true
#    declare -f tnew.title >/dev/null 2>&1                                     #= true
#    declare -f tnew.run   >/dev/null 2>&1                                     #= true
#    declare -f tnew.free  >/dev/null 2>&1                                     #= true
#    tnew.free
#  SOURCE
SYS::RSYNC() {

  local f fn s o=${1:-RSYNC}

  throw on InvalidVariable $o

  CFG ${o}_settings

  std::temp fn path="${TMPDIR}/${USER}/bashlyk" prefix='sys.rsync' suffix=".${o}"

  f=$( declare -f SYS::RSYNC::settings 2>/dev/null ) || error MissingMethod throw -- SYS::RSYNC::settings for $o
  for s in $_bashlyk_settings_sys_rsync; do

    echo "${f/SYS::RSYNC::settings/${o}.$s}" >> $fn  || error BadMethod     throw -- SYS::RSYNC::settings for $o

  done

  for s in $_bashlyk_methods_sys_rsync; do

    f=$( declare -f SYS::RSYNC::${s} 2>/dev/null ) || error MissingMethod throw -- SYS::RSYNC::${s} for $o
    echo "${f/SYS::RSYNC::$s/${o}.$s}" >> $fn      || error BadMethod     throw -- SYS::RSYNC::${s} for $o

  done

  source $fn || error InvalidArgument throw $fn
  return 0

}
#******
#****e* libsys/SYS::RSYNC::free
#  SYNOPSIS
#    SYS::RSYNC::free
#  DESCRIPTION
#    destructor of the instance
#  NOTES
#    public method
#  EXAMPLE
#    local i o s
#    SYS::RSYNC tRsync
#    tRsync.title = rsync instance test
#    tRsync.free                                                                 #? true
#    declare -f tRsync.title                                                     #? false
#  SOURCE
SYS::RSYNC::free() {

  local o s

  o=${FUNCNAME[0]%%.*}

  ${o}_settings.free

  for s in $_bashlyk_settings_sys_rsync $_bashlyk_methods_sys_rsync; do

    unset -f ${o}.$s

  done

}
#******
#****p* libsys/SYS::RSYNC::settings
#  SYNOPSIS
#    SYS::RSYNC::settings [ <value> ]
#  DESCRIPTION
#    set or get propertie(s) of the SYS::RSYNC instance
#  ARGUMENTS
#    <value> - set new value for selected properties
#              default, show property value
#  NOTES
#    fully virtual function, replaced by "get/set" methods when initializing an
#    instance:
#
#    title       - title message
#    options     - rsync options
#    pathSource  - source
#    pathTarget  - destination
#    truncateLog - max error lines to show
#    debugLevel  - debug level
#    onFailure   - run command on rsync fail
#    onSuccess   - run command on rsync success
#    fileLog     - name of a temporary file to save rsync stdout (may be readonly)
#  EXAMPLE
#    SYS::RSYNC tSettings
#    tSettings.title       = value
#    tSettings.pathSource  = /tmp/source
#    tSettings.pathTarget  = /tmp/target
#    tSettings.options     = --aCrv --delete-after
#    tSettings.debugLevel  = 2
#    tSettings.truncateLog = 10
#    tSettings.title                                                            | {{ '^value$'     }}
#    tSettings.pathSource                                                       | {{ ^/tmp/source$ }}
#    tSettings.pathTarget                                                       | {{ ^/tmp/target$ }}
#    tSettings.free
#  SOURCE
SYS::RSYNC::settings() {

  local o=${FUNCNAME[0]%%.*} f=${FUNCNAME[0]#*.}
  [[ $* ]] && ${o}_settings.set $f $* || ${o}_settings.get $f

}
#******
#****e* libsys/SYS::RSYNC::run
#  SYNOPSIS
#    SYS::RSYNC::run
#  DESCRIPTION
#    run rsync configured by instance
#  NOTES
#    public method
#  ERRORS
#    child processes
#  TODO
#    improve verbosity level handling
#  EXAMPLE
#    mkdir -p /tmp/qwaszx1
#    date -R > /tmp/qwaszx1/qwaszx3.txt
#    SYS::RSYNC rsync
#    rsync.pathSource = /tmp/qwaszx1/
#    rsync.onSuccess  = rm -r /tmp/qwaszx1
#    rsync.pathTarget = /tmp/qwaszx2
#    rsync.run
#    test -f /tmp/qwaszx2/qwaszx3.txt                                           #? false
#    test -f /tmp/qwaszx2/qwaszx3.txt                                           #? true
#    rm -r /tmp/qwaszx{1,2}
#    rsync.free
#  SOURCE
SYS::RSYNC::run() {

  local I fnErr fnStd fnRC rc o s V
  local -a a

  o=${FUNCNAME[0]%%.*}
  I=$( ${o}.truncateLog )
  : ${DEBUGLEVEL:=$( ${o}.debugLevel )}

  err::debugf 1 '%s:>' "$( ${o}.title )"

  std::temp fnErr
  std::temp fnRC
  std::temp fnStd
  ${o}.fileLog = $fnStd

  while read; do

    err::debugf 1 '%s' '.'
    echo "$REPLY" >> $fnStd

  done< <( rsync $( ${o}.options ) $( ${o}.pathSource ) $( ${o}.pathTarget ) 2>$fnErr; echo $? >$fnRC )

  rc=$( < $fnRC )
  if (( rc > 0 )); then

    err::debug 1 'fail..'
    a=( $( wc -l $fnErr ) )

    (( ${a[0]} == 1 )) || err::debug 1 'warns:'

    if (( ${a[0]} > I )); then

      head -n $(( I/2 )) $fnErr && echo '...' && tail -n $(( I/2 )) $fnErr

    else

      (( ${a[0]} > 0 )) && std::cat < $fnErr

    fi >&2

    _bashlyk_hError[$rc]="${_rsync_hError[$rc]}"
    s="$( ${o}.onFailure )"

    if [[ $s =~ ^(echo|warn|return|echo\+return|warn\+return|exit|echo\+exit|warn\+exit|throw)$ ]]; then

      error $rc $s -- rsync: $( ${o}.pathSource ) - $( ${o}.pathTarget )

    else

      [[ $s ]] && eval "$s"

    fi

    return $rc

  else

    err::debug 1 'ok!'
    s="$( ${o}.onSuccess )" && eval "$s"
    return 0

  fi
}
#******

